include/crab/domains/mru_region_domain.hpp

```diff
@@ -0,0 +1,2270 @@
+        }
+      }
+      return out_csts;
+    }
+  }
+
+  // Convert the abstract state into a disjunction of conjunction
+  // of linear constraints.
+  disjunctive_linear_constraint_system_t
+  to_disjunctive_linear_constraint_system() const override {
+    CRAB_ERROR("mru_region_domain::to_disjunctive_linear_constraint_system not "
+               "implemented");
+  }
+
+  std::string domain_name() const override {
+    return "MRURegionDomain(" + m_cache_lines.domain_name() + ", " +
+           m_regs.domain_name() + ")";
+  }
+
+  void write(crab_os &o) const override {
+    if (is_bottom()) {
+      o << "_|_";
+    } else if (is_top()) {
+      o << "{}";
+    } else {
+      CRAB_LOG("mru-region-print", o << "Mem=\t" << m_mem << ",\n"
+                                     << "( CacheLine=\t" << m_cache_lines
+                                     << ",\n"
+                                     << "Regs=\t" << m_regs << ",\n"
+                                     << "Addrs=\t" << m_addrs << " )\n";);
+
+      std::unordered_map<std::string, std::string> renaming_map;
+      auto get_type_fn = [](const variable_t &v) -> variable_type {
+        return variable_type(variable_type_kind::UNK_TYPE);
+      };
+      // Assigns a string name to each ghost variable
+      ghost_variables_t::mk_renaming_map(m_rev_var_map, get_type_fn,
+                                         renaming_map);
+      m_alloc.add_renaming_map(renaming_map);
+      o << "Mem=\t" << m_mem << ",\n"
+        << "( CacheLine=\t" << m_cache_lines << ",\n"
+        << "Regs=\t" << m_regs << ",\n"
+        << "Addrs=\t" << m_addrs << " )\n";
+      m_alloc.clear_renaming_map();
+    }
+  }
+  /**------------- End domain API definitions -------------------**/
+
+  /**------------- Begin MRU API definitions -------------------**/
+
+  // The expand performs a copy of region vars from mem into cache_lines
+  //  i.e. by given a most recently used object o, we copy all regions with o
+  //  in mem into the cache:
+  // For instance, o used V1, V2:
+  //    mem: { 1 <= V1 <= 2; V1 < V2; ... }
+  // The cache line will be:
+  //    cache_lines: { 1 <= V1 <= 2; V1 < V2; }
+  void expand(const variable_t &rgn) {
+    variable_vector_t Vs_for_new_obj;
+    auto it_var = m_rev_obj_rgn_map.find(rgn);
+    if (it_var == m_rev_obj_rgn_map.end()) {
+      // this should not happen
+      CRAB_ERROR("mru_region_domain::expand the mru object should be found");
+    }
+    variable_t representative = it_var->second;
+    auto it_rev_vars = m_obj_rgn_map.find(representative);
+    if (it_rev_vars == m_obj_rgn_map.end()) {
+      // this should not happen
+      CRAB_ERROR("mru_region_domain::expand the mru object should be found");
+    }
+    Vs_for_new_obj = it_rev_vars->second;
+
+    // expand does not need to perform base variable renaming
+    mru_region_domain_t tmp = mru_region_domain_t(*this);
+    tmp.project(Vs_for_new_obj);
+
+    // perform expand
+    m_cache_lines |= tmp.m_mem;
```

Comment:
It should be a meet. The join happens in `fold` to simulate weak update.

Reply:
We are not renaming the region variables. Maybe it's not needed in the end and this is what you intend.